lib/OnyxCache.js

```diff
@@ -0,0 +1,148 @@
+import _ from 'underscore';
+import lodashMerge from 'lodash/merge';
+
+
+const isDefined = _.negate(_.isUndefined);
+
+/**
+ * In memory cache providing data by reference
+ * Encapsulates Onyx cache related functionality
+ */
+class OnyxCache {
+    constructor() {
+        /**
+         * @private
+         * Cache of all the storage keys available in persistent storage
+         * @type {Set<string>}
+         */
+        this.storageKeys = new Set();
+
+        /**
+         * @private
+         * A map of cached values
+         * @type {Record<string, *>}
+         */
+        this.storageMap = {};
+
+        /**
+         * @private
+         * Captured pending tasks for already running storage methods
+         * @type {Record<string, Promise>}
+         */
+        this.pendingPromises = {};
+
+        // bind all methods to prevent problems with `this`
+        _.bindAll(
+            this,
+            'getAllKeys', 'getValue', 'hasCacheForKey', 'addKey', 'set', 'remove', 'merge',
+            'hasPendingTask', 'getTaskPromise', 'captureTask',
+        );
+    }
+
+    /**
+     * Get all the storage keys
+     * @returns {string[]}
+     */
+    getAllKeys() {
+        return [...this.storageKeys];
```

Comment:
Any reason we can't just return `this.storageKeys` directly?

Reply:
This actually converts the keys to array
It can be changed to something like `Array.from(this.storageKeys)` to be more explicit about what's happening

We can return `this.storageKeys` but then all usages have to be checked if they still work with a `Set` as currently they expect an `Array` 

https://github.com/Expensify/react-native-onyx/blob/27c0ba8ba1dee20171062ff249d03f2696d078a8/lib/Onyx.js#L373-L382

E.g. I'm not sure if this _.filter will still work, and don't want to break something

Reply:
I like the idea of switching this  to `Array.from`, it's a bit more explicit and should hopefully prevent someone from potentially making the mistake I almost made here.

Reply:
```suggestion
        return Array.from(this.storageKeys);
```